### PR TITLE
Making PrintValues() and ToString() able to not print default elements

### LINKED ISSUE #287: Permit detecting an element / attribute was explicitly set (user) or implicitly set (conversion / default value)
From discussion on 2020-01-23 on Asana.

Notes:
Default values
* Know if a property was actually defined
* `libsdformat`, in its conversion phase, substitutes in default values
* Odd / non-strict value parsing (e.g. colors, RGBA)
* API: Could add an attribute check, `IsExplicitlySet()` or `UsedDefaultValue()` or something, and have overloads for querying that uses `optional<>` or returns a flag
* May be hard to check if a default value was actually set?
* Could be useful for deprecating attributes that a user set?

\cc @SeanCurtis-TRI

### PR DESCRIPTION
**Pull request by marcoag**
Related to #287

## Summary
Now PrintValues() and ToString() will not output the elements that were added by default during parsing. There's also the option to call them with a boolean flag that will allow the caller to decide weather they want defaults int the output or not.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] Code check passed (In source directory, run `sh tools/code_check.sh`)
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/sdf/Element.hh [hunk 1]
```diff
@@ -149,6 +149,12 @@ namespace sdf
     /// \param[in] _prefix String value to prefix to the output.
     public: void PrintValues(std::string _prefix) const;
 
+    /// \brief Output Element's values to stdout.
+    /// \param[in] _prefix String value to prefix to the output.
+    /// \param[in] _includeDefaults flag for default elements and attributes.
+    public: void PrintValues(std::string _prefix,
```
Comment (aaronchongth) [inline include/sdf/Element.hh:155]: We should pass in the `std::string` as a `const std::string &`

File: include/sdf/Element.hh [hunk 2]
```diff
@@ -149,6 +149,12 @@ namespace sdf
...
+                             const bool _includeDefaults) const;
```
Comment (aaronchongth) [inline include/sdf/Element.hh:156]: We won't need `const` when we are passing the parameter by value.

File: include/sdf/Element.hh [hunk 3]
```diff
@@ -171,6 +177,13 @@ namespace sdf
     /// \return The string representation.
     public: std::string ToString(const std::string &_prefix) const;
 
+    /// \brief Convert the element values to a string representation.
+    /// \param[in] _prefix String value to prefix to the output.
+    /// \param[in] _includeDefaults flag for default elements and attributes.
+    /// \return The string representation.
+    public: std::string ToString(const std::string &_prefix,
+                                 const bool _includeDefaults) const;
```
Comment (aaronchongth) [inline include/sdf/Element.hh:185]: We won't need `const` when we are pssing by value.

File: include/sdf/Element.hh [hunk 4]
```diff
@@ -424,13 +437,16 @@ namespace sdf
     /// \param[in] _prefix arbitrary prefix to put on the string.
     /// \param[out] _out the std::ostreamstream to write output to.
     private: void ToString(const std::string &_prefix,
-                           std::ostringstream &_out) const;
+                           std::ostringstream &_out,
+                           const bool _includeDefaults) const;
```
Comment (aaronchongth) [inline include/sdf/Element.hh:441]: We won't need `const` when we are pssing by value.

File: include/sdf/Element.hh [hunk 5]
```diff
@@ -424,13 +437,16 @@ namespace sdf
...
 
     /// \brief Generate a string (XML) representation of this object.
     /// \param[in] _prefix arbitrary prefix to put on the string.
+    /// \param[in] _printDefaults flag for default elements and attributes.
     /// \param[out] _out the std::ostreamstream to write output to.
     private: void PrintValuesImpl(const std::string &_prefix,
-                                  std::ostringstream &_out) const;
+                                  std::ostringstream &_out,
+                                  const bool _includeDefaults) const;
```
Comment (aaronchongth) [inline include/sdf/Element.hh:449]: We won't need `const` when we are pssing by value.

**Review by aaronchongth — CHANGES_REQUESTED** [review]
Just did a quick look through and requested some changes to the API functions. Looks great! Are all the bullet points in #287 addressed at this point, such that it closes it?

File: include/sdf/Element.hh (hunk 1 above)
Comment (marcoag) [reply include/sdf/Element.hh]: addressed on https://github.com/osrf/sdformat/pull/575/commits/c9780654c9ba09a09d250b5776141a0dcd428b80

**Comment by marcoag** [comment]
I thought we got to that point, but maybe @azeey should confirm...

**Comment by azeey** [comment]
> I thought we got to that point, but maybe @azeey should confirm...

I think #287 only needed #542, but I remember @marcoag mentioning there were still some issues when upconverting SDFormat files. If that's the case, we should probably keep it open until that's resolved.

**Review by aaronchongth — APPROVED** [review]
This looks awesome! Thanks for making the changes!

**Comment by azeey** [comment]
Per VC, we should keep the default behavior (when printing via `ign sdf -p` or using `sdf::Element::ToString()` of printing values and elements added by default for `libsdformat9-libsdformat11`. For `libsdformat12`, we'll change the default behavior to print only user added content.

**Comment by marcoag** [comment]
Now it's a bit of a weird behavior since it will print any default `Element` but not their default `Attributes`. Shall we add another flag to print or not attributes in a similar way to `Element` here?

**Comment by azeey** [comment]
I suggest we have two behaviors: (1) a default behavior that is consistent with what the code did before this PR, (2) a new behavior enabled by a flag that does **not** print default elements or attributes.

**Comment by marcoag** [comment]
As agreed on our previous VC this has been added: 
- [x] kept behavior consistent with the code before this PR (print/toString default elements and not print/toString default attributes). 
- [x] Added two flags: one to print/toString default elements and another one to print/toString default attributes

File: src/Element.cc [hunk 6]
```diff
@@ -442,47 +442,59 @@ void Element::PrintDocLeftPane(std::string &_html, int _spacing,
...
+         aiter != this->dataPtr->attributes.end(); ++aiter)
     {
-      (*eiter)->ToString(_prefix + "  ", _out);
+      // Only print attribute values if they were set
+      // TODO(anyone): GetRequired is added here to support up-conversions where
+      // a new required attribute with a default value is added. We would have
+      // better separation of concerns if the conversion process set the
+      // required attributes with their default values.
+      if ((*aiter)->GetSet() || (*aiter)->GetRequired() ||
+          _includeDefaultAttributes)
```
Comment (marcoag) [inline src/Element.cc:463]: Still unsure if we should do something about this comment here...

**Review by azeey — APPROVED** [review]
LGTM! Just a few minor comments.

File: src/Element_TEST.cc [hunk 7]
```diff
@@ -575,6 +574,71 @@ TEST(Element, ToStringInclude)
...
+  sdf::ElementPtr elem = std::make_shared<sdf::Element>();
+  elem->SetParent(parent);
+  parent->InsertElement(elem);
+  sdf::ElementPtr elem2 = std::make_shared<sdf::Element>();
+  elem2->SetParent(parent);
+  parent->InsertElement(elem2);
+
+  std::ostringstream stream;
+  stream
+    << "<>\n"
```
[the remark this answers is not in the document]
Comment (marcoag) [reply src/Element_TEST.cc]: Solved in https://github.com/osrf/sdformat/pull/575/commits/f90794ae9cbfb8d2aff768e7b01bd8ca3d541c37#.

File: src/Element_TEST.cc [hunk 8]
```diff
@@ -575,6 +574,71 @@ TEST(Element, ToStringInclude)
...
+{
+  sdf::ElementPtr element = std::make_shared<sdf::Element>();
+  element->AddAttribute("test", "string", "foo", false, "foo description");
+  element->AddAttribute("test2", "string", "bar", true, "bar description");
+
+  EXPECT_EQ(element->ToString(""), element->ToString("", true, false));
+  EXPECT_EQ(element->ToString(""), element->ToString("", false, false));
+
+  std::ostringstream stream;
+  stream << "< test2='bar'/>\n";
```
[the remark this answers is not in the document]
Comment (marcoag) [reply src/Element_TEST.cc]: Solved in https://github.com/osrf/sdformat/pull/575/commits/f90794ae9cbfb8d2aff768e7b01bd8ca3d541c37#.

File: include/sdf/Element.hh [hunk 9]
```diff
@@ -171,6 +179,15 @@ namespace sdf
     /// \return The string representation.
     public: std::string ToString(const std::string &_prefix) const;
 
+    /// \brief Convert the element values to a string representation.
```
[the remark this answers is not in the document]
Comment (marcoag) [reply include/sdf/Element.hh:182]: Added in https://github.com/osrf/sdformat/pull/575/commits/f90794ae9cbfb8d2aff768e7b01bd8ca3d541c37#.

**Comment by marcoag** [comment]
As per VC we agreed to not consider default elements those added when the auto-conversion to the latest sdf version happens.

**Pull request merged**